pkg/git/git.go

```diff
@@ -0,0 +1,118 @@
+type UtilsRepository interface {
+	Worktree() (*git.Worktree, error)
+	Push(o *git.PushOptions) error
+}
+
+// utilsGit interface abstraction of git to enable tests
+type utilsGit interface {
+	plainClone(path string, isBare bool, o *git.CloneOptions) (*git.Repository, error)
+}
+
+// abstractedGit abstraction of git to enable tests
+var abstractedGit utilsGit = abstractionGit{}
+
+// TheGitUtils structure to provide git utilities
+type TheGitUtils struct {
+}
+
+// CommitSingleFile Commits the file located in the relative file path with the commitMessage to the given worktree.
+// In case of errors, the error is returned. In the successful case the commit is provided.
+func (TheGitUtils) CommitSingleFile(filePath, commitMessage string, worktree UtilsWorkTree) (plumbing.Hash, error) {
+	_, gitAddError := worktree.Add(filePath)
+	if gitAddError != nil {
+		log.Entry().WithError(gitAddError).Error("Failed to add file to git")
+		return [20]byte{}, gitAddError
+	}
+
+	commit, commitError := worktree.Commit(commitMessage, &git.CommitOptions{})
+	if commitError != nil {
+		log.Entry().WithError(commitError).Error("Failed to commit file")
+		return [20]byte{}, commitError
+	}
+
+	return commit, nil
+}
+
+// PushChangesToRepository Pushes all committed changes in the repository to the remote repository
+func (TheGitUtils) PushChangesToRepository(username, password string, repository UtilsRepository) error {
+	pushOptions := &git.PushOptions{
+		Auth: &http.BasicAuth{Username: username, Password: password},
+	}
+	pushError := repository.Push(pushOptions)
+	if pushError != nil {
+		log.Entry().WithError(pushError).Error("Failed to push commit")
+		return pushError
+	}
+	return nil
+}
+
+// PlainClone Clones a non-bare repository to the provided directory
+func (TheGitUtils) PlainClone(username, password, serverURL, directory string) (UtilsRepository, error) {
+	gitCloneOptions := git.CloneOptions{
+		Auth: &http.BasicAuth{Username: username, Password: password},
+		URL:  serverURL,
+	}
+	repository, gitCloneError := abstractedGit.plainClone(directory, false, &gitCloneOptions)
+	if gitCloneError != nil {
+		log.Entry().WithError(gitCloneError).Error("Failed to clone git")
+		return nil, gitCloneError
+	}
+	return repository, nil
+}
+
+// ChangeBranch checkout the provided branch.
+// It will create a new branch if the branch does not exist yet.
+// It will checkout "master" if no branch name if provided
+func (TheGitUtils) ChangeBranch(branchName string, worktree UtilsWorkTree) error {
+	if branchName == "" {
+		branchName = "master"
+	}
+
+	var checkoutOptions = &git.CheckoutOptions{}
+	checkoutOptions.Branch = plumbing.NewBranchReferenceName(branchName)
+	checkoutOptions.Create = false
+	checkoutError := worktree.Checkout(checkoutOptions)
+	if checkoutError != nil {
+		// branch might not exist, try to create branch
+		checkoutOptions.Create = true
+		checkoutError = worktree.Checkout(checkoutOptions)
```

Comment:
I'm not sure about the fall-back here. What about a new repo where the main branch is called "main"? If you pass no branchName, this repo will now have a new branch "master".

Reply:
yes, but I did not found a way to retrieve the main development branch.

Reply:
I am not completely sure what my point is, but perhaps that this is misplaced in the package. Perhaps there should be no fallback, and instead an error is returned if `branchName == ""`. What do you think?